File: 프로그래머스/lv2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
def solution(str1, str2):
    str1 = str1.upper() + "."
    str2 = str2.upper() + "."
    l1 = []
    l2 = []
    d1 = {}
    d2 = {}

    for i in range(len(str1) - 1):
        if ('A' <= str1[i] <= 'Z') and ('A' <= str1[i + 1] <= 'Z'):
            tmp = str1[i] + str1[i + 1]
            l1.append(tmp)
    for i in range(len(str2) - 1):
        if ('A' <= str2[i] <= 'Z') and ('A' <= str2[i + 1] <= 'Z'):
            tmp = str2[i] + str2[i + 1]
            l2.append(tmp)

    for i in range(len(l1)):
        if l1[i] in d1:
            d1[l1[i]] += 1
        else:
            d1[l1[i]] = 1
    for i in range(len(l2)):
        if l2[i] in d2:
            d2[l2[i]] += 1
        else:
            d2[l2[i]] = 1
    l1 = list(set(l1))
    l2 = list(set(l2))
    s = 0  # sum
    a = 0  # and
    for i in range(len(l1)):
        if l1[i] in d1 and l1[i] in d2:
            a += min(d2[l1[i]], d1[l1[i]])
            s += max(d2[l1[i]], d1[l1[i]])
            del d1[l1[i]]
            del d2[l1[i]]
        elif l1[i] in d1 and l1[i] not in d2:
            s += d1[l1[i]]
            del d1[l1[i]]
    for i in range(len(l2)):
        if l2[i] in d2 and l2[i] in d1:
            a += min(d2[l2[i]], d1[l2[i]])
            s += max(d2[l2[i]], d1[l2[i]])
            del d1[l2[i]]
            del d2[l2[i]]
        elif l2[i] in d2 and l2[i] not in d1:
            s += d2[l2[i]]
            del d2[l2[i]]
    if s == 0:
        return 65536
    else:
        answer = int((a / s) * 65536)
    return answer

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.java
import java.util.*;

/*
    둘 다 해쉬맵에 넣어놓고, 겹치는 것 중 적은 것의 개수가 교집합의 개수,
    겹치는 것 중 많은 것의 수가 합칩합의 개수,, 
    
    A먼저 사악 보고, 겹치는 것 A에서도 삭제 B에서도 삭제 
    B는 보면서, 합집합의 개수만 늘어남 
    
    또한, 대문자빼고는 버린다
    만약 둘다 공집합일면? 
*/

class Solution {
    public int solution(String str1, String str2) {
        int answer = 0;
        
        str1 = str1.toUpperCase();
        str2 = str2.toUpperCase();
        
        HashMap<String, Integer> hm1 = new HashMap<>();
        HashMap<String, Integer> hm2 = new HashMap<>();
        
        for (int i = 0; i < str1.length()-1; i++) {
            if (str1.charAt(i) < 'A' || str1.charAt(i) > 'Z') continue;
            if (str1.charAt(i+1) < 'A' || str1.charAt(i+1) > 'Z') continue; 
            
            hm1.put(str1.substring(i, i+2), hm1.getOrDefault(str1.substring(i, i+2), 0) + 1);
        }

        for (int i = 0; i < str2.length()-1; i++) {
            if (str2.charAt(i) < 'A' || str2.charAt(i) > 'Z') continue;
            if (str2.charAt(i+1) < 'A' || str2.charAt(i+1) > 'Z') continue; 
            
            hm2.put(str2.substring(i, i+2), hm2.getOrDefault(str2.substring(i, i+2), 0) + 1);
        }
        
        if (hm1.size() == 0 && hm2.size() == 0) {
            return 65536;
        }
        
        int a = 0;
        int b = 0;
                
        for (String hm1Key : hm1.keySet()) {
            a += Math.min(hm1.get(hm1Key), hm2.getOrDefault(hm1Key, 0));
            b += Math.max(hm1.get(hm1Key), hm2.getOrDefault(hm1Key, 0));
            
            hm1.put(hm1Key, 0);
            
            if (hm2.containsKey(hm1Key)) {
                hm2.put(hm1Key, 0);    
            }
        }
        
        for (String hm2Key : hm2.keySet()) {
            a += Math.min(hm2.get(hm2Key), hm1.getOrDefault(hm2Key, 0));
            b += Math.max(hm2.get(hm2Key), hm1.getOrDefault(hm2Key, 0));
            
            hm2.put(hm2Key, 0);
            
            if (hm1.containsKey(hm2Key)) {
                hm1.put(hm2Key, 0);    
            }
        }
        
        return a*65536/b;
    }
}